Student/Student/FilePicker/FilePickerViewController.swift

```diff
@@ -45,177 +43,225 @@ class FilePickerViewController: UIViewController, FilePickerViewProtocol {
-        }
+    @IBAction
+    func addFromFiles(_ sender: UIButton) {
+        let documentTypes = utis.map { $0.rawValue }
+        let documentPicker = UIDocumentPickerViewController(documentTypes: documentTypes, in: .import)
+        documentPicker.delegate = self
+        present(documentPicker, animated: true, completion: nil)
     }
 
-    private func toggleProgressBar(show: Bool) {
-        show ? showProgressBar() : hideProgressBar()
+    @IBAction
+    func addFromAudioRecorder(_ sender: UIButton) {
+        AudioRecorderViewController.requestPermission { allowed in
+            if allowed {
+                let controller = AudioRecorderViewController.create()
+                // controller.delegate = self
+                let nav = UINavigationController(rootViewController: controller)
+                self.present(nav, animated: true, completion: nil)
+            } else {
+                self.showPermissionError(.microphone)
+            }
+        }
     }
 
-    private func showProgressBar() {
+    func showProgressBar() {
         self.view.layoutIfNeeded()
         UIView.animate(withDuration: 0.4) {
             self.progressView?.isHidden = false
             self.view.layoutIfNeeded()
         }
     }
 
-    private func hideProgressBar() {
+    func hideProgressBar() {
         self.view.layoutIfNeeded()
         UIView.animate(withDuration: 0.4) {
             self.progressView?.isHidden = true
             self.view.layoutIfNeeded()
         }
     }
 
-    func updateToolbar(items: [UIBarButtonItem]) {
-        DispatchQueue.main.async {
-            self.setToolbarItems(items, animated: true)
-            self.navigationController?.setToolbarHidden(items.isEmpty, animated: true)
-            self.view.setNeedsLayout()
-            self.showSources(items.isEmpty)
-        }
-    }
-
-    func dismiss() {
-        DispatchQueue.main.async {
-            self.dismiss(animated: true, completion: nil)
-        }
-    }
-
-    func showSources(_ shown: Bool) {
-        sourcesView.isHidden = !shown
-        dividerView.isHidden = !shown
+    func hideMediaRecorder() {
     }
 }
 
 extension FilePickerViewController: UIDocumentPickerDelegate {
     func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
         for url in urls {
-            presenter?.add(fromURL: url)
+            delegate?.add(self, url: url)
         }
     }
 }
 
 extension FilePickerViewController: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
     public func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) {
         picker.dismiss(animated: true, completion: nil)
-        presenter?.add(withCameraResult: info)
+        do {
+            if let image = info[.originalImage] as? UIImage {
+                delegate?.add(self, url: try image.write())
```

Comment:
Do we not want the `.editedImage`? Or could we just work with the `.imageURL`?

Reply:
Oh I forget about `.editedImage`. I can check for that then fallback to `.originalImage`.

Reply:
Looks like `.imageURL` is always the `originalImage` url. I think it's good to try for edited image. Good call.